sys/include/net/ng_udp.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_net_udp UDP
+ * @ingroup     net
+ * @brief       RIOT's implementation of the UDP protocol
+ *
+ * @{
+ *
+ * @file
+ * @brief       UDP interface definition
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ */
+
+#ifndef NG_UDP_H_
+#define NG_UDP_H_
+
+#include <stdint.h>
+
+#include "byteorder.h"
+#include "net/ng_netbase.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Default message queue size for the UDP thread
+ */
+#ifndef NG_UDP_MSG_QUEUE_SIZE
+#define NG_UDP_MSG_QUEUE_SIZE   (8U)
+#endif
+
+/**
+ * @brief   Priority of the pktdump thread
+ */
+#ifndef NG_UDP_PRIO
+#define NG_UDP_PRIO             (PRIORITY_MAIN - 2)
+#endif
+
+/**
+ * @brief   Default stack size to use for the UDP thread
+ */
+#ifndef NG_UDP_STACK_SIZE
+#define NG_UDP_STACK_SIZE       (KERNEL_CONF_STACKSIZE_DEFAULT)
+#endif
+
+/**
+ * @brief   Size of a UDP header
+ */
+#define NG_UDP_HDR_LENGTH       (8U)
+
+/**
+ * @brief   UDP header definition
+ */
+typedef struct __attribute__((packed)) {
+    uint16_t src_port;      /**< source port, in network byte order */
+    uint16_t dst_port;      /**< destination port, network byte order */
+    uint16_t length;        /**< payload length (including the header),
+                             *   network byte order */
+    uint16_t checksum;      /**< checksum */
+} ng_udp_hdr_t;
+
+/**
+ * @brief   Calculate the checksum for the given packet
+ *
+ * @param[in] pkt       Pointer to the network layer header
+ *
+ * @return  0 on success
+ * @return  -EBADMSG if @p pkt does not contain the header data needed
+ * @return  -EFAULT if @p pkt is NULL
+ * @return  -ENOENT if @p pseudo_hdr_type is not known
+ */
+int ng_udp_calculate_csum(ng_pktsnip_t *hdr);
+
+/**
+ * @brief   Initialize and start UDP
+ *
+ * @return                  PID of the UDP thread
+ * @return                  negative value on error
```

Comment:
indention is fixed. I start to question the concept of very precise return values. In the past weeks it seemed to me like I am adding extra code (and thereby increasing code size), just to figure out which of 5 different error return values I should return. We might want to remember that we are programming (very) constrained systems and can maybe not have the most extensive error handling...